src/main/java/TaskList.java

```diff
@@ -0,0 +1,253 @@
+        }
+    }
+
+    public void markAsDone(String command) throws DukeException {
+        int taskIndex = extractTaskIndex(command);
+        if (taskIndex >= 0 && taskIndex < tasks.size()) {
+            tasks.get(taskIndex).markAsDone();
+            System.out.println(" Nice! I've marked this task as done:");
+            System.out.println("   " + tasks.get(taskIndex));
+        } else {
+            throw new DukeException("Invalid task index.");
+        }
+    }
+
+    public void markAsNotDone(String command) throws DukeException {
+        int taskIndex = extractTaskIndex(command);
+        if (taskIndex >= 0 && taskIndex < tasks.size()) {
+            tasks.get(taskIndex).markAsNotDone();
+            System.out.println(" OK, I've marked this task as not done yet:");
+            System.out.println("   " + tasks.get(taskIndex));
+        } else {
+            throw new DukeException("Invalid task index.");
+        }
+    }
+
+    private int extractTaskIndex(String command) throws DukeException {
+        String[] parts = command.split(" ");
+        if (parts.length < 2) {
+            throw new DukeException("OOPS!!! Please provide the task index.");
+        }
+        return Integer.parseInt(parts[1].trim()) - 1;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        if (!tasks.isEmpty()) {
+            for (int i = 0; i < tasks.size(); i++) {
+                sb.append(" ").append(i + 1).append(".").append(tasks.get(i));
+                sb.append("\n");
+            }
+        }
+        return sb.toString();
+    }
+
+    public void saveTasksToFile() {
+        try {
+            FileWriter writer = new FileWriter(filePath);
+            for (Task task : tasks) {
+                writer.write(task.toFileString() + "\n");
+            }
+            writer.close();
+        } catch (IOException e) {
+            System.out.println("Error saving tasks to file: " + e.getMessage());
+        }
+    }
+
+    public void loadTasksFromFile() {
+        try {
+            File file = new File(filePath);
+            if (!file.exists()) {
+                return; // If the file doesn't exist yet, no need to load tasks
+            }
+
+            BufferedReader reader = new BufferedReader(new FileReader(file));
+            String line;
+            while ((line = reader.readLine()) != null) {
+                processFileLine(line);
+            }
+            reader.close();
+        } catch (IOException | DukeException e) {
+            System.out.println("Error loading tasks from file: " + e.getMessage());
+        }
+    }
+
+    private void processFileLine(String line) throws DukeException {
+        // Parse the line and create tasks based on the format in the file
+        String[] parts = line.split(" \\| ");
+        if (parts.length != 3) {
+            throw new DukeException("Invalid task format in the file.");
```

Comment:
Perhaps subtypes of DukeException can be defined to specify error message printed instead of manually entering the error message every time DukeExceptions are thrown.